content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md

```diff
@@ -0,0 +1,141 @@
++++
+type = "blog"
+title = "Debugging RISCV-64 bootloader in QEMU"
+author = "kallisti5"
+date = "2021-01-17 09:36:04-06:00"
+tags = ["haiku", "software", "riscv64", "qemu"]
++++
+
+Debugging early bootloader code can be extremely difficult. The lack of printf and other classical black-box debugging tools means you're limited to one of the following methods of debugging lockups:
+
+1. Arm chair debugging, Changing code, compiling, booting, repeat.
+2. Using GDB or another debugger tool and stepping through code watching outcome.
+
+Obviously #2 above is more ideal than #1.  Welcome to debugging by attaching gdb to qemu!
+
+The steps below really should apply to any architecture. Just replace riscv64 with your architecture of choice.
+
+## Pre-requirements
+
+* All of the Haiku compiling [pre-requisite software](/guides/building/pre-reqs) installed.
+* ArchLinux and Fedora offer riscv64 toolchains.
+  * I recommend installing them. (ex: riscv64-elf-gdb, riscv64-elf-gcc, riscv64-elf-binutils)
+
+### Compiling Haiku (with debug symbols)
+
+Overall, this is the easier step for most folks with any Haiku development experience.
+
+1. Get the [haiku and buildtools git repos](/guides/building/get-source-git).
+2. Compile your toolchain.
+   1. ```cd haiku; mkdir generated.riscv64; cd generated.riscv64```
+   1. ```../configure -j2 --cross-tools-source ../../buildtools --build-cross-tools riscv64```
+3. Enable debugging symbols in bootloader.
+   1. Copy UserBuildConfig.example to UserBuildConfig in haiku/build/jam/
```

Comment:
That's not really needed, you can start from an empty file and just add a single SetConfigVar line to it.

Reply:
fair.  I just normally use it as an "example"